tests/mobilenet_tests.cpp

```diff
@@ -0,0 +1,147 @@
+  model.Predict(input, output);
+  REQUIRE(arma::accu(output) == Approx(singleBatchOutput).epsilon(1e-2));
+
+  // Run prediction for multiple batch.
+  model.Predict(multipleBatchInput, output);
+  REQUIRE(arma::accu(output) == Approx(multipleBatchOutput).epsilon(1e-2));
+}
+
+// General ANN serialization test.
+template<typename LayerType>
+void ModelSerializationTest(LayerType& layer)
+{
+  arma::mat input(224 * 224 * 3, 10, arma::fill::randu);
+  arma::mat output(1000, 10, arma::fill::randu);
+
+  FFN<> model;
+  model.Add<LayerType>(layer);
+
+  model.InputDimensions() = std::vector<size_t>({224, 224, 3});
+
+  // Takes only one pass over the input data.
+  ens::StandardSGD opt(0.1, 1, 5, -100, false);
+  model.Train(input, output, opt);
+
+  arma::mat originalOutput;
+  model.Predict(input, originalOutput);
+
+  // Now serialize the model.
+  FFN<> xmlModel, jsonModel, binaryModel;
+  SerializeObjectAll(model, xmlModel, jsonModel, binaryModel);
+
+  // Ensure that predictions are the same.
+  arma::mat modelOutput, xmlOutput, jsonOutput, binaryOutput;
+  model.Predict(input, modelOutput);
+  xmlModel.Predict(input, xmlOutput);
+  jsonModel.Predict(input, jsonOutput);
+  binaryModel.Predict(input, binaryOutput);
+
+  CheckMatrices(originalOutput, modelOutput, 1e-5);
+  CheckMatrices(originalOutput, xmlOutput, 1e-5);
+  CheckMatrices(originalOutput, jsonOutput, 1e-5);
+  CheckMatrices(originalOutput, binaryOutput, 1e-5);
+}
+
+TEST_CASE("MobileNetSerializationTest", "[MobileNetTests]")
+{
+  models::Mobilenet model;
+  ModelSerializationTest(model);
+}
+
+TEST_CASE("MobileNetTest", "[MobileNetTests]")
+{
+  arma::mat input(224 * 224 * 3, 1, arma::fill::randu);
+  arma::mat output;
+  models::Mobilenet mobilenetLayer;
+  FFN<> model;
+  model.InputDimensions() = std::vector<size_t>({224, 224, 3});
+  model.Add<models::Mobilenet>(mobilenetLayer);
+  ModelDimTest(model, input);
+}
+
+TEST_CASE("MobileNetMultiBatchTest", "[MobileNetTests]")
+{
+  arma::mat input(224 * 224 * 3, 10, arma::fill::randu);
+  arma::mat output;
+  models::Mobilenet mobilenetLayer;
+  FFN<> model;
+  model.InputDimensions() = std::vector<size_t>({224, 224, 3});
+  model.Add<models::Mobilenet>(mobilenetLayer);
+  ModelDimTest(model, input, 1000, 10);
+}
+
+TEST_CASE("MobileNetCustomTest", "[MobileNetTests]")
+{
+  arma::mat input(224 * 224 * 3, 10, arma::fill::randu);
+  arma::mat output;
+  models::Mobilenet mobilenetLayer(512);
+  FFN<> model;
+  model.InputDimensions() = std::vector<size_t>({224, 224, 3});
+  model.Add<models::Mobilenet>(mobilenetLayer);
+  ModelDimTest(model, input, 512, 10);
+}
+
+TEST_CASE("MobileNetNoTopTest", "[MobileNetTests]")
+{
+  arma::mat input(224 * 224 * 3, 10, arma::fill::randu);
+  arma::mat output;
+  models::Mobilenet mobilenetLayer(512, false);
+  FFN<> model;
+  model.InputDimensions() = std::vector<size_t>({224, 224, 3});
+  model.Add<models::Mobilenet>(mobilenetLayer);
+  ModelDimTest(model, input, 9216, 10);
+}
```

Comment:
It would be really great to add a test for `models::MobileNetType<arma::fmat>`, just to make sure that the model works with low precision.  I think so long as you don't test serialization, that should work just fine.

(The reason serialization would have a problem is because you'd need to register all the `arma::fmat` layers for serialization.  That can add a lot of compile time... but I think it's safe to say that if MobileNet serializes correctly with `arma::mat`, then it will serialize correctly with `arma::fmat`, assuming that each layer correctly serializes with `arma::fmat`.  From the perspective of this code, that should be a safe assumption---testing individual layer serialization should be done inside of mlpack's core code.)